MRG: Allow sphere='eeglab' in topomap plotting functions

hoechenberger opened the pull request:
This is based on @mmagnuski's example,
https://mne.tools/stable/auto_examples/visualization/eeglab_head_sphere.html

This is just a "dump" of work I'd done a few days ago because the default MNE topomaps for EEG led to confusion when I discussed the figures with my colleagues. There were still issues with the approach used here, depending on which montage was used … but I didn't have enough time to really investigate. In any case, these are my WIP results.

I won't have time to get back to this for a few days, so if anybody would like to take over, please do go ahead!

cc @cbrnr @sappelhoff

cc @agramfort @larsoner

mmagnuski commented:
Cool, I like the idea! Would this still work if one did not have these channels in the data? To be more specific, because I may be not up to date with how montages work now, is it possible to add positions of the required channels without associated data, and this approach would still work?

mmagnuski commented:
BTW, I can test/explore my question later in the evening if this is not known niw. But a tip on how to add positions of the required channels (without associated data) when one has the montage applied already would be usedul! :)

hoechenberger commented:
@mmagnuski If i remember correctly, at first I accidentally used a recording  without T7/T8 and there it didn't work, so I added the explicit checks here ... I believe when setting the montage, the DigPoints without corresponding electrodes were just dropped automatically:/ but maybe there's a switch to control that? And if not, maybe we should add one :)

mmagnuski commented:
> DigPoints without corresponding electrodes were just dropped automatically:/ but maybe there's a switch to control that? And if not, maybe we should add one :)

Yes, that would be useful. I think Eric @larsoner proposed something related at one point - so that the topomap sphere could be fit to all channels even when some were dropped.

larsoner commented:
https://github.com/mne-tools/mne-python/issues/8195

hoechenberger commented:
I'd say we can already move forward here before #8195 is resolved -- as that appears to be a not-so-easy adjustment...

hoechenberger commented:
@mmagnuski Can I hand this over to you? 😅 or maybe to @sappelhoff? As I'm currently very busy with other stuff :(

hoechenberger commented on mne/utils/check.py:
```diff
@@ -815,12 +815,48 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+            y = pos[-1, 1]
+            z = pos[:, -1].mean()
+
+            sphere=(x, y, z, radius)
```
```suggestion
            sphere = (x, y, z, radius)
```

hoechenberger commented on mne/utils/check.py:
```diff
@@ -815,12 +815,48 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+
+            sphere=(x, y, z, radius)
+            sphere_units = 'm'
+            del x, y , z, radius, montage, ch_pos
```
```suggestion
            del x, y, z, radius, montage, ch_pos
```

mmagnuski commented:
@hoechenberger I can take over once I finish plot_adjacency and the other PR. But by that time you will have likely finished the PR already ;)

hoechenberger commented:
Nope, feel free to leave this one on your agenda 😅 mine is exploding 😱🤓

hoechenberger commented:
@mmagnuski Do you fancy picking this one up again? I think it would be a great addition to 1.1 :)

hoechenberger commented on mne/utils/check.py:
```diff
@@ -846,13 +846,87 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+            sphere = tuple(r0) + (R,)
+            sphere_units = 'm'
+        elif sphere == 'eeglab':
+            # We need coordinates for for the 2D plane formed by
```
```suggestion
            # We need coordinates for the 2D plane formed by
```

hoechenberger commented on mne/utils/check.py:
```diff
@@ -846,13 +846,87 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+            montage = info.get_montage()
+            if montage is None:
+                raise ValueError(
+                    'No montage was found on your data, but sphere="eeglab" '
```
```suggestion
                    'No montage was set on your data, but sphere="eeglab" '
```

hoechenberger commented:
This is ready for review.

The code demands that `info['dig']` be set, which may not be the case for older datasets, where we **could** rely on the channel locations in `info['chs'][...]['loc']` instead. But trying to do that felt like using an old and dated API somehow, so I decided to only support going through the digpoints. Also, users could simply use `set_montage()` to create those, so I didn't feel like going the extra mile to support an (outdated?) use case.

I've reworked the channel locations tutorial to make it easier to discover this new functionality; and I've removed @mmagnuski's example demonstrating how to produce "EEGLAB-style" topoplots manually.


Screenshot from the tutorial:
<img width="500" alt="Screen Shot 2022-06-09 at 21 28 28" src="https://user-images.githubusercontent.com/2046265/172928907-1c15d07f-80ba-45ec-bc9a-48eb707c265f.png">


Rendered tutorial:
https://output.circle-artifacts.com/output/job/c091c2cf-b479-4702-9092-7d4670e5de7e/artifacts/0/dev/auto_tutorials/intro/40_sensor_locations.html

cc @drammock @agramfort

hoechenberger commented:
Latest revision of the rendered tutorial:

https://output.circle-artifacts.com/output/job/33450733-1333-437a-9fe0-85186671cf77/artifacts/0/dev/auto_tutorials/intro/40_sensor_locations.html#sphx-glr-auto-tutorials-intro-40-sensor-locations-py

agramfort reviewed (approved):
@drammock feel free to merge.

🙏 @hoechenberger

hoechenberger commented:
Latest tutorial render:
https://output.circle-artifacts.com/output/job/3791974e-36f1-4958-9beb-d3005f2c86bd/artifacts/0/dev/auto_tutorials/intro/40_sensor_locations.html#sphx-glr-auto-tutorials-intro-40-sensor-locations-py

Example API docs:
https://output.circle-artifacts.com/output/job/3791974e-36f1-4958-9beb-d3005f2c86bd/artifacts/0/dev/generated/mne.channels.DigMontage.html#mne.channels.DigMontage.plot

cbrnr commented:
Great work @hoechenberger! I don't have time to review this anytime soon, so please go ahead without me. The only thing I would like to remark is that it would be great to fix the "TP10" label being out of bounds in the tutorial. This will surely be confusing. Maybe @drammock knows what to do?

hoechenberger commented:
> The only thing I would like to remark is that it would be great to fix the "TP10" label being out of bounds in the tutorial. This will surely be confusing. Maybe @drammock knows what to do?

Yes that would be great. I just confirmed that when running the tutorial locally from the command line, the label displays correctly. It's really just the tutorial rendering via sphinx-gallery (?) that seems to be off.

hoechenberger commented on doc/changes/latest.inc:
```diff
@@ -89,7 +89,9 @@ Enhancements
-- :meth:`mne.Report.add_figure` and :meth:`mne.Report.add_image` gained a new parameter, ``section``, allowing to put multiple figures or images into the same "content block". Previously, the only way to group figures and images was by means of a slider, which sometimes made content hard to discover. The slider remains available as an alternative. (:gh:`10694` by `Richard Höchenberger`_)
+- :meth:`mne.Report.add_figure` and :meth:`mne.Report.add_image` gained a new parameter, ``section``, allowing to put multiple figures or images into the same "content block". Previously, the only way to group figures and images was by means of a slider, which sometimes made content hard to discover. The slider remains available as an alternative (:gh:`10694` by `Richard Höchenberger`_)
+
+- All functions and method that plot topographic maps for EEG (2D projections of the EEG sensor locations) now accept the parameter value ``sphere='eeglab'`` to lay out the sensors with respect to the head circle in a similar way to how EEGLAB does, i.e., T7/T8 and Fpz/Oz are placed directly on the circle for template montages (:gh:`10572` by `Richard Höchenberger`_)
```
```suggestion
- All functions and methods that plot topographic maps for EEG (2D projections of the EEG sensor locations) now accept the parameter value ``sphere='eeglab'`` to lay out the sensors with respect to the head circle in a similar way to how EEGLAB does, i.e., T7/T8 and Fpz/Oz are placed directly on the circle for template montages (:gh:`10572` by `Richard Höchenberger`_)
```

hoechenberger commented on mne/utils/check.py:
```diff
@@ -846,13 +846,87 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+                    'to apply a montage.'
+                )
+            ch_pos = montage.get_positions()['ch_pos']
+
```
```suggestion
```

hoechenberger commented on mne/utils/check.py:
```diff
@@ -846,13 +846,87 @@ def _check_sphere(sphere, info=None, sphere_units='m'):
+                        )
+                    raise ValueError(msg)
+
+            #  Calculate the radius from: T7<->T8, Fpz<->Oz
```
```suggestion
            # Calculate the radius from: T7<->T8, Fpz<->Oz
```

drammock commented on doc/glossary.rst:
```diff
@@ -369,14 +369,6 @@ general neuroimaging concepts. If you think a term is missing, please consider
         See :class:`SourceSpaces` for the class definition.
 
     stim channel
-        See :term:`trigger channel`.
```
@hoechenberger such cross-references are unnecessary; sphinx glossary allows multiple terms to have the same definition (see, e.g., `dipole, ECD, equivalent current dipole`) and for alphabetization of the entries the first term given is the one considered.

drammock commented on mne/viz/utils.py:
```diff
@@ -1133,6 +1133,15 @@ def _plot_sensors(pos, info, picks, colors, bads, ch_names, title, show_names,
+        extents = [x.get_window_extent(renderer=renderer) for x in ax.texts]
+        xmaxs = np.array([x.max[0] for x in extents])
+        bad_xmax_ixs = np.nonzero(xmaxs > xmax)[0]
+        needed_space = (xmaxs[bad_xmax_ixs] - xmax).max() / xmax
```
the readability of this line is a special gift for you, @hoechenberger :)

hoechenberger commented:
Thanks @drammock, I appreciate your effort! I can take a look at this tomorrow if you want feedback?

drammock commented:
ok, TP10 is now visible: https://output.circle-artifacts.com/output/job/08e3fbf6-9d3d-4b17-948d-30692e433408/artifacts/0/dev/auto_tutorials/intro/40_sensor_locations.html#plotting-2d-sensor-locations-like-eeglab

Docstring looks good too:
https://output.circle-artifacts.com/output/job/08e3fbf6-9d3d-4b17-948d-30692e433408/artifacts/0/dev/generated/mne.channels.DigMontage.html#mne.channels.DigMontage.plot

Once CIs are all green I'll fix the `sphinx_gallery_thumbnail_number` so it doesn't show the offset head as the thumbnail 🤦🏻 and then merge.

drammock commented:
thx @hoechenberger and @mmagnuski!
